discovery/azure/azure.go

```diff
@@ -380,6 +442,14 @@ func (client *azureClient) getVMs(ctx context.Context) ([]virtualMachine, error)
 	}
 	for result.NotDone() {
 		for _, vm := range result.Values() {
+			if vm.InstanceView == nil {
```

Comment:
It might be better to do this in the `for i, vm := range machines { ... }` loop as the calls to the Azure API will be done in parallel instead of sequentially here.

Reply:
Is there a way to do this without additional RPCs?

Azure has really low api limits, and so far we've had to reject all PRs that added more RPCs due to this.

Reply:
@simonpasquier We cannot do it inside `for i, vm := range machines { ... }` because `machines` contains both individual VMs and VMSS' VMs. There are different API calls to get InstanceView for each of the types.

@brian-brazil Unfortunately that's the way Azure has implemented it. I will keep monitoring the updates for azure-sdk-for-go, in case they would address this issue.

I've resolved the rest of the comments and updated the PR. Please take a look.

Reply:
This PR proposes to roughly double the RPCs we send to Azure, which is a showstopper unfortunately as it halves the (already relatively small) number of machines that can be monitored. Is there any other way to get this, or cut other RPCs?

Reply:
@brian-brazil 
> This PR proposes to roughly double the RPCs we send to Azure
> Is there any other way to get this, or cut other RPCs?

To be precise, before this PR we were making one API call `result, err := client.vm.ListAll(ctx)` to get the list of all VMs.
In this PR I added iterating over the VMs and extracting InstanceView for every one of them.
I agree that this is not optimal, but `ListAll` does not populate InstanceView.
We have to make a separate API call for each VM.
There is an open issue in azure-sdk https://github.com/Azure/azure-sdk-for-python/issues/573
Azure team is supposed to resolve this problem at some point, but I don't know when.

I would suggest that we accept this PR, as it resolves a bug and provides additional metrics.
I will follow up with the azure team regarding their plans to fix this problem, and I will open a new PR once it's ready.

Reply:
We also need a call per machine to get network stats currently.

Reply:
> We also need a call per machine to get network addresses currently.

Correct, I'm aware of that.
But there is currently no other way to get power state (and disregard deallocated VMs) without sending a separate InstanceView request for every VM.

Reply:
CloudNativeCon is this week, hopefully I can find someone with Azure to have a chat about this. This is far from the first PR we've had to reject on these grounds :(

Reply:
Could you elaborate your concern about azure low API limits?

According to this doc (https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-manager-request-limits), we can make 12,000 read API calls per hour.

The discovery runs every 5 minutes, or 12 times in an hour.
Assuming we have 100 VMs in the cluster, we would be changing the number of read API calls from 12*100=1200 to 12*2*100=2400.
Unless a customer has over 500 VMs, we are not hitting the limit.
What is the typical/maximal Azure cluster size, in your experience?

Reply:
500 VMs is not that many, and 5m delay is already quite a lot. 1m is what we use for other SDs. When we initially chose those values we thought there were 15k requests per hour.

Reply:
I totally understand.
The only alternative I'm aware of is the Azure Instance Metadata Service, but it is available on the VM itself, not from a remote endpoint.
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service

Reply:
Had a chat with someone at the Azure stand. They didn't have anyone from that particular team there, but will put them in touch with me.

Reply:
hi @brian-brazil , any update on this?
My team govern around 5500 VMs in Azure and located in 90+ subscriotions.
I'm thinking of having one additional REST API call to get power state of VM is fine.